File: Games/Sevens/src/main/java/jokatu/game/games/sevens/input/CardInputDeserialiser.java
package jokatu.game.games.sevens.input;

import jokatu.game.input.DeserialisationException;
import jokatu.game.input.TypedSingleKeyInputDeserialiser;
import org.jetbrains.annotations.NotNull;
import org.springframework.stereotype.Component;

import java.util.Map;

import static jokatu.game.cards.Cards.CARDS_BY_DISPLAY;

@Component
public class CardInputDeserialiser extends TypedSingleKeyInputDeserialiser<String, CardInput> {

	@NotNull
	@Override
	public CardInput deserialiseTypedSingleValue(@NotNull Map<String, Object> json, @NotNull String cardDisplay) throws DeserialisationException {
		if (!CARDS_BY_DISPLAY.containsKey(cardDisplay)) {
			throw new DeserialisationException(json, "The value for 'card' was not a valid card.");
		}
		return new CardInput(CARDS_BY_DISPLAY.get(cardDisplay));
	}

	@NotNull
	@Override
	protected Class<String> getType() {
		return String.class;
	}

	@NotNull
	@Override
	protected String getKeyName() {
		return "card";
	}
}

File: Games/Sevens/src/main/kotlin/jokatu/game/games/sevens/input/CardInputDeserialiser.kt
package jokatu.game.games.sevens.input

import jokatu.game.cards.Cards
import jokatu.game.input.DeserialisationException
import jokatu.game.input.TypedSingleKeyInputDeserialiser
import org.springframework.stereotype.Component

@Component
class CardInputDeserialiser : TypedSingleKeyInputDeserialiser<String, CardInput>(String::class) {

	override val keyName: String
		get() = "card"

	@Throws(DeserialisationException::class)
	override fun deserialiseTypedSingleValue(json: Map<String, Any>, cardDisplay: String): CardInput {
		if (!Cards.CARDS_BY_DISPLAY.containsKey(cardDisplay)) {
			throw DeserialisationException(json, "The value for 'card' was not a valid card.")
		}
		return CardInput(Cards.CARDS_BY_DISPLAY[cardDisplay]!!)
	}
}
